presto-tablestore/src/main/java/com/facebook/presto/tablestore/TablestoreAbstractRecordCursor.java

```diff
@@ -0,0 +1,214 @@
+    protected final long startTimestamp;
+    protected final String tableName;
+    protected final ConnectorSession session;
+    protected final TablestoreFacade tablestoreFacade;
+    protected final TablestoreSplit split;
+    protected final TablestoreTableHandle tableHandle;
+    protected final boolean enableLooseCast;
+
+    protected boolean closed;
+    protected int rows;
+    protected long apiNano;
+    protected long readNano;
+
+    public TablestoreAbstractRecordCursor(TablestoreFacade tablestoreFacade, ConnectorSession session, TablestoreSplit split,
+            List<TablestoreColumnHandle> columnHandles)
+    {
+        this.tablestoreFacade = tablestoreFacade;
+        this.session = requireNonNull(session, "session is null");
+        this.columnHandles = requireNonNull(columnHandles, "columnHandles is null")
+                .toArray(new TablestoreColumnHandle[columnHandles.size()]);
+        this.split = split;
+        this.tableHandle = split.getTableHandle();
+        this.tableName = requireNonNull(tableHandle.getTableName(), "tableName is null");
+        this.primaryKeys = columnHandles.stream()
+                .filter(TablestoreColumnHandle::isPrimaryKey)
+                .sorted(Comparator.comparingInt(TablestoreColumnHandle::getPkPosition))
+                .collect(Collectors.toList());
+        this.enableLooseCast = TablestoreFacade.enableLooseCast(session);
+        this.startTimestamp = currentTimeMillis();
+    }
+
+    protected <T> T callAndRecord(int field, Callable<T> call)
+    {
+        long t1 = nanoTime();
+        try {
+            checkStates(field);
+            return call.call();
+        }
+        catch (Throwable t) {
+            throw new RuntimeException("row[pk=" + currentPrimaryKeyAsDebug() + "] field[index=" + field + "] -> " + t.getMessage(), t);
+        }
+        finally {
+            readNano += nanoTime() - t1;
+        }
+    }
+
+    @Override
+    public long getReadTimeNanos()
+    {
+        checkClosed();
+        return (currentTimeMillis() - startTimestamp) * NANOS_PER_MILLI_SECOND;
+    }
+
+    protected long getTotalCostMillis()
+    {
+        return getReadTimeNanos() / NANOS_PER_MILLI_SECOND;
+    }
+
+    protected long getApiCostMillis()
+    {
+        return apiNano / NANOS_PER_MILLI_SECOND;
+    }
+
+    protected long getReadCostMillis()
+    {
+        return readNano / NANOS_PER_MILLI_SECOND;
+    }
+
+    @Override
+    public Type getType(int field)
+    {
+        checkStates(field);
+        return columnHandles[field].getColumnType();
+    }
+
+    protected String getColumnName(int idx)
+    {
+        return columnHandles[idx].getColumnName();
+    }
+
+    protected boolean isPrimaryKey(int idx)
+    {
+        return columnHandles[idx].isPrimaryKey();
+    }
+
+    protected abstract boolean hasNext();
+
+    protected abstract void next();
+
+    protected abstract String currentPrimaryKeyAsDebug();
```

Comment:
shall we have a better name? hard to understand `currentPrimaryKeyAsDebug`

Reply:
It is used to print the current primary key(type and value), mainly for debug purpose when there is something wrong, renamed to currentPrimaryKeyInfo()